### PR TITLE
rustdoc: Turn `next_def_id` comments into docs

### PR DESCRIPTION
**Pull request by camelid**
Split out from #80740.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -291,7 +291,9 @@ impl Item {
         }
     }
 
-    /// See comments on next_def_id
+    /// See the documentation for [`next_def_id()`].
+    ///
+    /// [`next_def_id()`]: crate::core::DocContext::next_def_id()
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:296]:
```suggestion
    /// [`next_def_id()`]: DocContext::next_def_id()
```

File: src/librustdoc/core.rs [hunk 2]
```diff
@@ -120,14 +120,20 @@ impl<'tcx> DocContext<'tcx> {
...
-    // registered after the AST is constructed would require storing the defid mapping in a
-    // RefCell, decreasing the performance for normal compilation for very little gain.
-    //
-    // Instead, we construct 'fake' def ids, which start immediately after the last DefId.
-    // In the Debug impl for clean::Item, we explicitly check for fake
-    // def ids, as we'll end up with a panic if we use the DefId Debug impl for fake DefIds
+    /// Create a new "fake" [`DefId`].
+    ///
+    /// This is an ugly hack, but it's the simplest way to handle synthetic impls without greatly
+    /// refactoring either `librustdoc` or [`rustc_middle`]. In particular, allowing new [`DefId`]s
```
Comment (jyn514) [inline src/librustdoc/core.rs:126]: I don't think the directory is relevant here.
```suggestion
    /// refactoring either rustdoc or [`rustc_middle`]. In particular, allowing new [`DefId`]s
```

**Review by jyn514 — APPROVED** [review]

File: src/librustdoc/core.rs (hunk 2 above)
Comment (camelid) [reply src/librustdoc/core.rs:126]: Well `librustdoc` is also the name of the crate, but yeah I think it should be changed.

Comment (jyn514) [reply src/librustdoc/core.rs]: The name of the crate is `rustdoc`. The name of the binary is `rustdoc-tool`.

**Pull request merged**